### ISSUE TITLE
Can't mix Protocol and Abstract class

### ISSUE THREAD

**Issue by twoertwein** [beartype/beartype#117]
The following code roars but I think it should purr :)
```py
from abc import ABC, abstractmethod
from typing import Protocol, runtime_checkable

from beartype import beartype

@runtime_checkable
class Test(Protocol):
    something: int

class AbstractTest(Test, ABC):
    something = 1

    def foo(self):
        ...

    @abstractmethod
    def bar(self):
        ...

class NotAbstractTest(AbstractTest):
    def bar(self):
        ...

@beartype
def test(x: AbstractTest) -> None:
    ...

test(NotAbstractTest())
```

> beartype.roar.BeartypeDecorHintPep484Exception: PEP 484 generic <class '__main__.AbstractTest'> method resolution order (<class '__main__.AbstractTest'>, <class '__main__.Test'>, <class 'typing.Protocol'>, <class 'typing.Generic'>, <class 'abc.ABC'>, <class 'object'>) second-to-last item not <class 'typing.Generic'>.

**Comment by leycec** [comment]
> The following code roars but I think it should purr...

**So, it comes to this.** My own userbase uses my puns against me. I see I now have no recourse but to:

1. Reluctantly roll up my greasy sleeves.
1. Squash another shameful bug in the `beartype 0.10.x` series (which I will subsequently pretend never existed).
1. Excrete a new `beartype 0.10.5` point release with this fix.
1. Troll everyone with an even more un*bear*able bear pun.

*We begin.*

**Comment by twoertwein** [comment]
Thank you for working on it! The initial example might not be very realistic (there shouldn't be a reason to have a non-abstract&non-protocol class inherit from a Protocol). Here is a similar example:


```py
from abc import ABC, abstractmethod
from typing import Protocol, runtime_checkable

from beartype import beartype

@runtime_checkable
class Test(Protocol):
    something: int

class Abstract(ABC):
    @abstractmethod
    def bar(self):
        ...

class AbstractTest(Test, Abstract):
    ...

class NotAbstract(Abstract):
    something = 1

    def bar(self):
        ...

@beartype
def test(x: AbstractTest) -> None:
    ...

test(NotAbstract())
```

**Comment by leycec** [comment]
> Thank you for working on it!

*Au contraire*. Thank **you** for uncovering this horrorshow! I never would have thought to combine two things that should not be combined – but you technically can, so we technically better support that.

@beartype: the codebase that supports impossible paradoxes.

> ...probably good not to mix abstract and Protocol at all

Are you psychic? You might be. I'd monetize that, but it might get risky when the government finds out. When they do, **I do not know Torsten Wörtwein.**

Now that we've clarified that for your handlers, it's probably pertinent to note that **all protocols are abstract base classes (ABCs).** Yes, it's true! Internally, the metaclass of `typing.Protocol` is `typing._ProtocolMeta(abc.ABCMeta)`. Externally, protocols are *usually* defined as fully abstract using the exact same boilerplate `abc` decorators we all know and abuse everywhere.

For example, here's the standard `typing.SupportsInt` protocol:

```python
@runtime_checkable
class SupportsInt(Protocol):
    """An ABC with one abstract method __int__."""
    __slots__ = ()

    @abstractmethod  # <-------- it's happening
    def __int__(self) -> int:
        pass
```

For that reason, it's thankfully unnecessary to explicitly subclass protocols from `abc.ABC`. I mean, you technically *can* – I think. But you certainly don't need to and probably shouldn't. Of course, I'm all for defying assumptions by anachronistically smashing everything with a hammer.

Barking arctic pup seal approves. :seal:

**Comment by leycec** [comment]
**Bwa-hah-_hah_.** Resolved by 11bc17f, whose resolution is best depicted by this accurate rendition of my life for the past 10 commits.

![it do be like that](https://cdn.kapwing.com/collections/20-minute-adventure-rick-morty-meme-template-9hzf6.jpg)

Things rapidly spiralled out of control after I miserably realized I'd never *really* tested caching `beartype.typing.Protocol`-based protocols. Now they are... but it cost us all something to get there.

**Oh!** And while we're on the subject, consider using caching `beartype.typing.Protocol`-based protocols rather than non-caching `typing.Protocol`-based protocols. They're *much* faster. Also, you no longer need to explicitly decorate everything with `@typing.runtime_checkable`. We now do that for you.

This is the power of `beartype.typing`, because you deserve the best:

```python
from abc import ABC, abstractmethod
from beartype.typing import Protocol  # <-- *THIS IS THE WAY*
from beartype import beartype

class Test(Protocol):
    something: int

class Abstract(ABC):
    @abstractmethod
    def bar(self):
        ...

class AbstractTest(Test, Abstract):
    ...

class NotAbstract(Abstract):
    something = 1

    def bar(self):
        ...

@beartype
def test(x: AbstractTest) -> None:
    ...

test(NotAbstract())
``` 

Thanks again for the excellent issue, @twoertwein. Let's try not to do this again. :face_exhaling:
